arwen/elrondapi/managedei.go

```diff
@@ -0,0 +1,894 @@
+	blockchain := arwen.GetBlockchainContext(context)
+	managedType := arwen.GetManagedTypesContext(context)
+
+	gasToUse := metering.GasSchedule().ElrondAPICost.GetExternalBalance
+	metering.UseGas(gasToUse)
+
+	address, err := managedType.GetBytes(addressHandle)
+	if err != nil {
+		_ = arwen.WithFault(arwen.ErrArgOutOfRange, context, runtime.ElrondAPIErrorShouldFailExecution())
+		return
+	}
+	tokenID, err := managedType.GetBytes(tokenIDHandle)
+	if err != nil {
+		_ = arwen.WithFault(arwen.ErrArgOutOfRange, context, runtime.ElrondAPIErrorShouldFailExecution())
+		return
+	}
+
+	esdtToken, err := blockchain.GetESDTToken(address, tokenID, uint64(nonce))
+	if err != nil {
+		_ = arwen.WithFault(arwen.ErrArgOutOfRange, context, runtime.ElrondAPIErrorShouldFailExecution())
+		return
+	}
+
+	value := managedType.GetBigIntOrCreate(valueHandle)
+	value.Set(esdtToken.Value)
+
+	managedType.SetBytes(propertiesHandle, esdtToken.Properties)
+	if esdtToken.TokenMetaData != nil {
+		managedType.SetBytes(hashHandle, esdtToken.TokenMetaData.Hash)
+		managedType.ConsumeGasForThisIntNumberOfBytes(len(esdtToken.TokenMetaData.Hash))
+		managedType.SetBytes(nameHandle, esdtToken.TokenMetaData.Name)
+		managedType.ConsumeGasForThisIntNumberOfBytes(len(esdtToken.TokenMetaData.Name))
+		managedType.SetBytes(attributesHandle, esdtToken.TokenMetaData.Attributes)
+		managedType.ConsumeGasForThisIntNumberOfBytes(len(esdtToken.TokenMetaData.Attributes))
+		managedType.SetBytes(creatorHandle, esdtToken.TokenMetaData.Creator)
+		managedType.ConsumeGasForThisIntNumberOfBytes(len(esdtToken.TokenMetaData.Creator))
+		royalties := managedType.GetBigIntOrCreate(royaltiesHandle)
+		royalties.SetUint64(uint64(esdtToken.TokenMetaData.Royalties))
+
+		totalBytes := writeManagedVecOfManagedBuffers(managedType, esdtToken.TokenMetaData.URIs, urisHandle)
+		managedType.ConsumeGasForThisIntNumberOfBytes(int(totalBytes))
+	}
+}
+
+//export v1_4_managedAsyncCall
+func v1_4_managedAsyncCall(context unsafe.Pointer, destHandle int32, valueHandle int32, dataHandle int32) {
+	host := arwen.GetVMHost(context)
+	runtime := host.Runtime()
+	metering := host.Metering()
+	managedType := host.ManagedTypes()
+
+	gasSchedule := metering.GasSchedule()
+	gasToUse := gasSchedule.ElrondAPICost.AsyncCallStep
+	metering.UseGas(gasToUse)
+
+	address, err := managedType.GetBytes(destHandle)
+	if err != nil {
+		_ = arwen.WithFault(arwen.ErrArgOutOfRange, context, runtime.ElrondAPIErrorShouldFailExecution())
+		return
+	}
+
+	data, err := managedType.GetBytes(dataHandle)
+	if err != nil {
+		_ = arwen.WithFault(arwen.ErrArgOutOfRange, context, runtime.ElrondAPIErrorShouldFailExecution())
+		return
+	}
+
+	value, err := managedType.GetBigInt(valueHandle)
+	if err != nil {
+		_ = arwen.WithFault(arwen.ErrArgOutOfRange, context, runtime.ElrondAPIErrorShouldFailExecution())
+		return
+	}
+
+	gasToUse = math.MulUint64(gasSchedule.BaseOperationCost.DataCopyPerByte, uint64(len(data)))
+	metering.UseGas(gasToUse)
```

Comment:
Use this gas immediately after `data, err := managedType.GetBytes(dataHandle)`

Consider creating a new method `managedType.GetBytesMetered(handle)` which automatically consumes gas.